libdnf/base/goal.cpp

```diff
@@ -1469,21 +1474,46 @@ void Goal::Impl::add_group_remove_to_goal(
                     continue;
                 }
 
-                // if the package is required by another installed package, it is
-                // not removed, but it's reason is changed to DEPENDENCY
-                rpm::PackageQuery dependent(query_installed);
-                dependent.filter_requires(pkg.get_provides());
-                if (dependent.size() > 0) {
-                    rpm_goal.add_reason_change(pkg, transaction::TransactionItemReason::DEPENDENCY, std::nullopt);
-                    continue;
-                }
+                remove_candidates.emplace_back(std::move(pkg));
+            }
+            rpm_goal.add_group(group, transaction::TransactionItemAction::REMOVE, reason);
+        }
+    }
+    if (remove_candidates.empty()) {
+        return;
+    }
 
-                add_remove_to_goal(transaction, pkg_name, pkg_settings);
+    // all installed packages, that are not candidates for removal
+    rpm::PackageQuery dependent_base(query_installed);
+    {
+        // create auxiliary goal to resolve all unused dependencies that are going to be
+        // removed together with removal candidates
+        Goal goal_tmp(base);
+        for (const auto & pkg : remove_candidates) {
+            goal_tmp.add_rpm_remove(pkg);
+        }
+        for (const auto & tspkg : goal_tmp.resolve().get_transaction_packages()) {
+            if (transaction_item_action_is_outbound(tspkg.get_action())) {
+                dependent_base.remove(tspkg.get_package());
             }
+        }
+    }
 
-            rpm_goal.add_group(group, transaction::TransactionItemAction::REMOVE, reason);
+    // The second step of packages removal - filter out packages that are
+    // dependencies of a package that is not also being removed.
+    libdnf::solv::IdQueue packages_to_remove_ids;
+    for (const auto & pkg : remove_candidates) {
+        // if the package is required by another installed package, it is
+        // not removed, but it's reason is changed to DEPENDENCY
+        rpm::PackageQuery dependent(dependent_base);
+        dependent.filter_requires(pkg.get_provides());
+        if (dependent.size() > 0) {
+            rpm_goal.add_reason_change(pkg, transaction::TransactionItemReason::DEPENDENCY, std::nullopt);
+            continue;
         }
+        packages_to_remove_ids.push_back(pkg.get_id().id);
     }
+    rpm_goal.add_remove(packages_to_remove_ids, cfg_main.clean_requirements_on_remove().get_value());
```

Comment:
Hmm this is interesting, I didn't think of it before. Because all tests are passing, everything is green, but I suppose after merging it as it is, it wouldn't be compilable. Can we improve that with like running the CI stack in the environment where the commit is already rebased and merged?

Reply:
I think that already happens, the problem is it ran before merging the incompatible changes. I believe if we re-run the CI now it would fail to compile.

Reply:
I have tried to re-run it just now to make sure.

Reply:
Allright, then just maybe automatic re-running workflows for pending pull requests? But that's probably expensive... At least there could be some warning that the checks are not valid anymore, it is confusing :slightly_smiling_face: